log4j-api/src/main/java/org/apache/logging/log4j/internal/map/UnmodifiableArrayBackedMap.java

```diff
@@ -350,82 +351,59 @@ public UnmodifiableArrayBackedMap copyAndRemove(String key) {
+                : StreamSupport.stream(keysToRemoveIterable.spliterator(), false)
+                        .collect(Collectors.toSet());
 
-        int firstIndexToKeep = -1;
-        int lastIndexToKeep = -1;
-        int destinationIndex = 0;
-        int numEntriesKept = 0;
-        // build the new map
-        UnmodifiableArrayBackedMap newMap = new UnmodifiableArrayBackedMap(numEntries);
-        for (int indexInCurrentMap = 0; indexInCurrentMap < numEntries; indexInCurrentMap++) {
-            // for each key in this map, check whether it's in the set we built above
-            Object key = backingArray[getArrayIndexForKey(indexInCurrentMap)];
-            if (!keysToRemoveSet.contains(key)) {
-                // this key should be kept
-                if (firstIndexToKeep == -1) {
-                    firstIndexToKeep = indexInCurrentMap;
-                }
-                lastIndexToKeep = indexInCurrentMap;
-            } else if (lastIndexToKeep > 0) {
-                // we hit a remove, copy any keys that are known ready
-                int numEntriesToCopy = lastIndexToKeep - firstIndexToKeep + 1;
-                System.arraycopy(
-                        backingArray,
-                        getArrayIndexForKey(firstIndexToKeep),
-                        newMap.backingArray,
-                        getArrayIndexForKey(destinationIndex),
-                        numEntriesToCopy * 2);
-                firstIndexToKeep = -1;
-                lastIndexToKeep = -1;
-                destinationIndex += numEntriesToCopy;
-                numEntriesKept += numEntriesToCopy;
-            }
-        }
+        // Create the new map
+        final UnmodifiableArrayBackedMap oldMap = this;
+        final int oldMapEntryCount = oldMap.numEntries;
+        final UnmodifiableArrayBackedMap newMap = new UnmodifiableArrayBackedMap(oldMapEntryCount);
 
-        if (lastIndexToKeep > -1) {
-            // at least one key still requires copying
-            int numEntriesToCopy = lastIndexToKeep - firstIndexToKeep + 1;
-            System.arraycopy(
-                    backingArray,
-                    getArrayIndexForKey(firstIndexToKeep),
-                    newMap.backingArray,
-                    getArrayIndexForKey(destinationIndex),
-                    numEntriesToCopy * 2);
-            numEntriesKept += numEntriesToCopy;
+        // Short-circuit if there is nothing to remove
+        if (keysToRemove.isEmpty()) {
+            System.arraycopy(oldMap.backingArray, 0, newMap.backingArray, 0, oldMapEntryCount * 2);
+            newMap.numEntries = oldMapEntryCount;
+            return this;
         }
 
-        if (numEntriesKept == 0) {
-            return EMPTY_MAP;
+        // Iterate over old map entries
+        int newMapEntryIndex = 0;
+        for (int oldMapEntryIndex = 0; oldMapEntryIndex < oldMapEntryCount; oldMapEntryIndex++) {
+            final int oldMapKeyIndex = getArrayIndexForKey(oldMapEntryIndex);
```

Comment:
Much cleaner than the original, thanks.

Reply:
@vy,

Can you point out, where was the bug in the old implementation? It is not obvious to me.

Reply:
Added a dedicated test for the `copyAndRemoveAll()` in df5ec8115974970b5625836d9c06a7d72a65b813. The bug was on this line:

https://github.com/apache/logging-log4j2/blob/2d61fe0e6a3b9c8f5d53dda1f5834752a2799138/log4j-api/src/main/java/org/apache/logging/log4j/internal/map/UnmodifiableArrayBackedMap.java#L394

It skips the removal if the slice-to-be-copied is `[0,0]`.